modules/admin/src/test/java/org/motechproject/admin/it/AdminBundleIT.java

```diff
@@ -114,6 +129,56 @@ public void testMessageController() throws IOException, InterruptedException {
         assertTrue("No messages listed", json.size() > 0);
     }
 
+    @Test
+    public void testUploadBundleFromRepository() throws IOException, InterruptedException {
+        boolean isUploadSuccessful = uploadBundle("Repository", "org.motechproject:cms-lite:LATEST", null, "on");
+
+        assertTrue(isUploadSuccessful);
+    }
+
+    @Test
+    public void testUploadBundleFromFile() throws IOException, InterruptedException {
+        File file = new File("src/test/resources/motech-upload-test-bundle.jar");
+
+        boolean isUploadSuccessful = uploadBundle("File", null, file, "on");
+
+        assertTrue(isUploadSuccessful);
+    }
+
+    private boolean uploadBundle(String moduleSource, String moduleId, File bundleFile, String startBundle) throws IOException, InterruptedException {
+        String uri = String.format("http://%s:%d/admin/api/bundles/upload/", HOST, PORT);
+
+        HttpPost httpPost = new HttpPost(uri);
+        MultipartEntityBuilder entity = MultipartEntityBuilder.create();
+        Charset chars = Charset.forName("UTF-8");
+        entity.setCharset(chars);
+        entity.addTextBody("moduleSource", moduleSource, ContentType.MULTIPART_FORM_DATA);
+        if (moduleSource.equals("Repository")) entity.addTextBody("moduleId", moduleId, ContentType.MULTIPART_FORM_DATA);
+        else if (moduleSource.equals("File")) {
+            if (bundleFile == null) {
+                return false;
+            }
+            entity.addBinaryBody("bundleFile", bundleFile);
+        }
+        else return false;
+        entity.addTextBody("startBundle", startBundle, ContentType.MULTIPART_FORM_DATA);
+
+        httpPost.setEntity(entity.build());
+
+        int bundlesCountBeforeUpload = bundleContext.getBundles().length;
+        HttpResponse response = getHttpClient().execute(httpPost);
+        EntityUtils.consume(response.getEntity());
+        int bundlesCountAfterUpload = bundleContext.getBundles().length;
+
+        if (response.getStatusLine().getStatusCode() != 200) {
+            return false;
+        }
+
+        int bundleState = bundleContext.getBundles()[bundlesCountAfterUpload-1].getState();
```

Comment:
Is the order of the returned bundles guaranteed by the framework? If not, we should find the bundle by symbolic name.